File: 22.10_PS/1018_샘터.py
# boj. 18513

import sys
from collections import deque

input = sys.stdin.readline
n, k = map(int, input().split())
arr = list(map(int, input().split()))
count, answer = 0, 0
visited_dict = {}

queue = deque()
for a in arr:
    queue.append(a)
    visited_dict[a] = 0

while queue:
    q = queue.popleft()
    for x in [-1, 1]:
        if visited_dict.get(q + x) == None:
            visited_dict[q + x] = visited_dict[q] + 1
            answer += visited_dict[q] + 1
            queue.append(q + x)
            count += 1
        if count >= k: 
            sys.exit(print(answer))

File: 22.10_PS/1018_샘터.java
// boj. 18513

import java.io.*;
import java.util.*;

public class Main {
    static int[] arr;
    static HashMap<Integer, Integer> visited;
    static Queue<Integer> queue;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        arr = new int[n];
        queue = new LinkedList<>();
        visited = new HashMap<>();

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            queue.add(arr[i]);
            visited.put(arr[i], 0);
        }

        System.out.println(BFS(k));
    }

    static long BFS(int k) {
        long result = 0;
        int count = 0;
        while (!queue.isEmpty()) {
            int q = queue.poll();
            for (int i = -1; i < 2; i += 2) {
                if (!visited.containsKey(q + i)) {
                    visited.put((q + i), (visited.get(q) + 1));
                    result += visited.get(q + i);
                    queue.add(q + i);
                    count++;
                }
                if (count == k)
                    return result;
            }
        }
        return result;
    }
}